### PR TITLE
Updated API docs to sync with supported data types in csiwe_op_*.

### PR DESCRIPTION
**Pull request by hunkim**
Based on excellent comments from @yaroslavvb and @zheng-xq, this PR changes API docs to sync with supported data types in csiwe_op_*.

This addresses #2225

### CODE REVIEW

**Comment by martinwicke** [comment]
@girving Will this be a backward-compatibilty issue?

**Comment by girving** [comment]
It will break the backwards compatibility test, since it reduces the list of datatypes the ops are registered for.  However, it's not otherwise a backwards compatibility issue if the removed ops never ran, and it's a good change to make since it stops us from lying.  We should arguably make an exception and do the necessary surgery to fix the test.

I believe the backwards compatibility test is open source, so @hunkim: am I correct that you haven't tried running the unit tests yet with y our change?

**Comment by girving** [comment]
@josh11b: If you agree, what's the best way to perform the necessary surgery to clarify that `tf.sqrt` has never worked on `int32`?

**Comment by hunkim** [comment]
@girving, @martinwicke Not at all. Now this API doc is synced with actual implementation.

**Comment by martinwicke** [comment]
This is not a doc file. This is the file that actually defined which
kernels are compiled.

**Comment by girving** [comment]
@hunkim: I understand that your change is good, but by "Not at all" do you mean that you definitely haven't run the tests?  I believe they will fail, since we have a backwards compatibility check that registered ops only get more permissive (ignoring the kernels).

**Comment by hunkim** [comment]
@girving I just generated API docs and it was successful. So I assumed it was OK. It was my bad.

I just ran the python tests. They all passed, but as you said //tensorflow/python:nn_test failed. 

`TypeError: DataType int32 for attr 'T' not in list of allowed values: float16, float32, float64, complex64`

How to update docs without breaking this? Currently, supported types in doc and code are different as discussed PR #2227.

> A Tensor. Must be one of the following types: float32, float64, int32, complex64, int64

`
REGISTER4(UnaryOp, CPU, "Cos", functor::cos, float, Eigen::half, double, complex64);`

File: tensorflow/core/ops/math_ops.cc [hunk 1]
```diff
@@ -131,90 +131,102 @@ tf.complex_abs(x) ==> [5.25594902, 6.60492229]
 
 // Declares cwise unary operations signature: 't -> 't
 #define UNARY()                      \
-  Input("x: T").Output("y: T").Attr( \
-      "T: {half, float, double, int32, complex64, int64}")
+  Input("x: T").Output("y: T")
```
Comment (zheng-xq) [inline tensorflow/core/ops/math_ops.cc]: I suggest to add another macro such as UNARY_FP around the list of types. So it is easier to add types in the future. 

{half, float, double, complex64}

**Comment by zheng-xq** [comment]
Adding @girving to give guidance on how to deal with the backward compatibility test. 

In general, it is a good idea to run the TensorFlow tests locally.

**Comment by hunkim** [comment]
@zheng-xq Thanks for the tip. I'll always run tests locally even for doc changes.

However, testing takes a really long time for me: about 40 min in my Mac, and over 10 min in a good google cloud machine (16 vCPUs, 60 GB memory) even if I change only one code line.

I wonder why it runs all tests again and again for just one-line-change? Can it only run affected test cases by the change?

**Comment by zheng-xq** [comment]
>> I wonder why it runs all tests again and again for just one-line-change? Can it only run affected test cases by the change?

The public bazel does not have the capability to achieve that automatically. 

When you get to the point that you know the test structure, you can just run the tests that are most likely to fail.

File: tensorflow/core/ops/math_ops.cc (hunk 1 above)
Comment (hunkim) [inline tensorflow/core/ops/math_ops.cc]: Good idea. However, there is one more kind:

```
REGISTER_OP("Lgamma")
    .UNARY()
    .Attr("T: {half, float, double}")
    .Doc(R"doc(
Computes the log of the absolute value of `Gamma(x)` element-wise.
)doc");
```
Should we have three MACROS in total?

- UNARY:  {half, float, double, int32, complex64, int64}
- UNARY_FP:  {half, float, double, int64}
- UNARY_FPC:  {half, float, double, complex64}

I think it's getting complicated. For somehow, just adding `int32, int64` in the cwise_op* might be simpler.

` REGISTER4(UnaryOp, CPU, "Cos", functor::cos, float, Eigen::half, double, int32, int64, complex64);
`
 It won't add much as we discussed, but does it hurt? I guess there is no backwards compatibility issue, and no need to add two more macros.

Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: We definitely don't want trigonometric functions of integers.  I think the macros should be

* `UNARY`: Explicit list of types (what you did)
* `UNARY_ALL`: the current list of types
* `UNARY_REAL`: half, float, double (not sure why you stuck int64 in their)
* `UNARY_COMPLEX`: half, float, double, complex64, complex128

where the last three macros are implemented in terms of the first.

**Comment by girving** [comment]
Verifying with @josh11b, but I think we should do the surgery to make the backwards compatibility test pass.  However, you seem to say above that the backwards compatibility test doesn't fail with your change, which is suspicious.  To confirm, can you do

    bazel test core/ops/compat:backwards_compatibility_test

That should fail.  Assuming it does, it should be reasonably easy to fix by manually editing

     core/ops/compat/ops_history.v0.pbtxt

to remove the types that are invalid.  I will then look at the changes to that file very carefully. :)

File: tensorflow/core/ops/math_ops.cc (hunk 1 above)
Comment (hunkim) [inline tensorflow/core/ops/math_ops.cc]: Sounds good. 
" (not sure why you stuck int64 in their)" was a typo.

**Comment by girving** [comment]
@hunkim: Go for stripping the bad dtypes out of `ops_history.v0.pbtxt`.  Let me know if you have questions or want help.

File: tensorflow/core/ops/math_ops.cc (hunk 1 above)
Comment (hunkim) [inline tensorflow/core/ops/math_ops.cc]: @girving two quick questions for you:

- UNARY: {half, float, double, int32, complex64, int64}
- UNARY_ALL: {half, float, double, uint8, int8, int16, int32, int64, complex64}

Something like this? I am not sure UNARY_ALL is necessary.

- UNARY_REAL: {half, float, double}
- UNARY_COMPLEX: {half, float, double, complex64, ~~complex128~~}

I don't see `complex128` used in math_ops or cwise_op_\*? Delete `complex128` or add it in the MACRO and cwise_op_\*?

Thanks!

Comment (girving) [inline tensorflow/core/ops/math_ops.cc]: Since you're touching the code, let's go ahead and add `complex128` support.  It should just mean adding it to the macro here and a few register macros in `cwise_op_*` as you say.  Thanks!

Your list of macros seems good, except that we should go ahead and allow `complex128` everywhere that `complex64` is allowed.

Comment (hunkim) [inline tensorflow/core/ops/math_ops.cc]: Thanks! Then, let me submit two separate PRs:

1. adding the `complex128` support
2. adding 'UNARY' MACROS

for easy review and testing. I'm working on it.

Comment (hunkim) [inline tensorflow/core/ops/math_ops.cc]: For 1, I have submitted PR #2261. Thanks!

**Comment by hunkim** [comment]
I have submitted a new PR #2435.
